cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/AllocationToInvalidPointer.qll

```diff
@@ -124,45 +127,83 @@ module SizeBarrier {
     )
   }
 
+  /**
+   * Holds if `left < nRight + k` holds if `g` evaluates to `edge`.
+   */
   private predicate operandGuardChecks(
-    IRGuardCondition g, Operand left, Operand right, int state, boolean edge
+    IRGuardCondition g, Operand left, DataFlow::Node right, int k, boolean edge
   ) {
-    exists(DataFlow::Node nLeft, DataFlow::Node nRight, int k |
-      nRight.asOperand() = right and
-      nLeft.asOperand() = left and
-      SizeBarrierConfig::isSink(nLeft, nRight, g, k, edge) and
-      state = getAFlowStateForNode(nRight) and
-      k <= state
+    flowTo(right) and
+    SizeBarrierConfig::isSink(DataFlow::operandNode(left), right, g, k, edge)
+  }
+
+  /**
+   * Gets an instruction `instr` that is guarded by a check such as `instr <= left + delta` where
+   * `left <= _ + k` and `left` is the "small side" of of a relational comparison that checks
+   * whether `left <= size` where `size` is the size of an allocation.
+   */
+  Instruction getABarrierInstruction0(int delta, int k) {
+    exists(IRGuardCondition g, ValueNumber value, Operand left, boolean edge, DataFlow::Node right |
+      left = value.getAUse() and
+      // We know:
+      // 1. result <= value + delta (by `bounded`)
+      // 2. value < right + k + 1 (by `operandGuardChecks`).
+      // Condition 2 implies: value <= right + k, so if we know
+      // that `state >= k + delta` then we have:
```

Comment:
What is `state`?

Reply:
The `getABarrierInstruction0` predicate is called in the body of `getABarrierInstruction` which has a `state` parameter that the final bound we produce in `pointerAddInstructionHasBounds` relates to. In both places that call `getABarrierInstruction0` there's a `state` parameter (i.e., `getABarrierInstruction` and `getABarrierNode`), and the `state` in the comment here is supposed to refer to that `state`.

Thinking more about this, it may be better to simply remove any mention of `state` in this predicate and push those comments to the callers that actually have a `state` parameter 🤔.

Reply:
b750016bc1a6a01731b6e2d12d675ed07f633858 moves some comments around as I suggested above. Is that more clear?

Reply:
Yes, thanks.